server/src/main/java/io/deephaven/server/barrage/BarrageMessageProducer.java

```diff
@@ -715,23 +728,71 @@ private void enqueueUpdate(final TableUpdate upstream) {
-                        final long currKeyEnd =
-                                inverter.mapToPrevKeyspace(rowSet.get(Math.min(posEnd, rowSet.size() - 1)), true);
+                        final long currKeyStart, currKeyEnd;
+                        if (sub.reverseViewport) {
+                            // using the reverse ShiftHelper, must pass `key` in descending order
+                            currKeyEnd =
+                                    inverter.mapToPrevKeyspace(rowSet.get(Math.min(localEnd, rowSet.size() - 1)), true);
+                            currKeyStart =
+                                    inverter.mapToPrevKeyspace(rowSet.get(Math.min(localStart, rowSet.size() - 1)),
+                                            false);
+                        } else {
+                            // using the forward ShiftHelper, must pass `key` in ascending order
+                            currKeyStart =
+                                    inverter.mapToPrevKeyspace(rowSet.get(Math.min(localStart, rowSet.size() - 1)),
+                                            false);
+                            currKeyEnd =
+                                    inverter.mapToPrevKeyspace(rowSet.get(Math.min(localEnd, rowSet.size() - 1)), true);
+                        }
 
                         // if our current viewport includes no previous values this range may be empty
                         if (currKeyEnd < currKeyStart) {
                             return;
                         }
 
-                        final long prevKeyStart =
-                                posStart >= prevRowSet.size() ? prevRowSet.lastRowKey() + 1 : prevRowSet.get(posStart);
-                        final long prevKeyEnd = prevRowSet.get(Math.min(posEnd, prevRowSet.size() - 1));
+                        final long prevKeyStart, prevKeyEnd;
+
+                        if (sub.reverseViewport) {
+                            // compute positions to be relative to the final position of prevRowSet
+                            final long lastPrevRowPosition = prevRowSet.size() - 1;
+
+                            long prevStart = Math.max(lastPrevRowPosition - posEnd, 0);
+                            long prevEnd = lastPrevRowPosition - posStart;
+
+                            prevKeyStart =
+                                    prevStart >= prevRowSet.size() ? prevRowSet.lastRowKey() + 1
+                                            : prevRowSet.get(prevStart);
+                            prevKeyEnd = prevRowSet.get(Math.min(prevEnd, prevRowSet.size() - 1));
+                        } else {
+                            prevKeyStart =
+                                    localStart >= prevRowSet.size() ? prevRowSet.lastRowKey() + 1
+                                            : prevRowSet.get(posStart);
+                            prevKeyEnd = prevRowSet.get(Math.min(posEnd, prevRowSet.size() - 1));
+                        }
```

Comment:
```suggestion
                        final long prevStart;
                        final long prevEnd;
                        if (sub.reverseViewport) {
                            // compute positions to be relative to the final position of prevRowSet
                            final long lastPrevRowPosition = prevRowSet.size() - 1;

                            prevStart = Math.max(lastPrevRowPosition - posEnd, 0);
                            prevEnd = Math.max(lastPrevRowPosition - posStart, 0;
                        } else {
                            prevStart = localStart;
                            prevEnd = localEnd;
                        }
                        prevKeyStart =
                                prevStart >= prevRowSet.size() ? prevRowSet.lastRowKey() + 1  : prevRowSet.get(prevStart);
                        prevKeyEnd = prevRowSet.get(Math.min(prevEnd, prevRowSet.size() - 1));
```

Reply:
Interestingly, the suggested line `prevEnd = Math.max(lastPrevRowPosition - posStart, 0;` is incorrect or at least causes many test failures.  Investigating why this is happening.

Reply:
After digging in, it looks like passing a negative value into `prevRowSet.get()` returns `RowSequence.NULL_ROW_KEY` == `-1` but passing `0` returns the first row key.  `-1` will always trigger the `if (currKeyEnd > prevKeyEnd)` statement and add rows to the scope but the row key may not. 

@nbauernfeind Is this correct behavior?  Should we document this better or revise this to be clearer?